include/solvers/vof_scratch_data.h

```diff
@@ -0,0 +1,259 @@
+   *
+   * @param solution_stages The solution at the intermediary stages (for SDIRK methods)
+   *
+   * @param source_function The function describing the VOF source term
+   *
+   */
+
+  template <typename VectorType>
+  void
+  reinit(const typename DoFHandler<dim>::active_cell_iterator &cell,
+         const VectorType &                                    current_solution,
+         const std::vector<VectorType> &previous_solutions,
+         const std::vector<VectorType> &solution_stages)
+  {
+    this->fe_values_fs.reinit(cell);
+    quadrature_points = this->fe_values_fs.get_quadrature_points();
+    auto &fe_fs       = this->fe_values_fs.get_fe();
+
+    if (dim == 2)
+      this->cell_size = std::sqrt(4. * cell->measure() / M_PI) / fe_fs.degree;
+    else if (dim == 3)
+      this->cell_size = pow(6 * cell->measure() / M_PI, 1. / 3.) / fe_fs.degree;
+
+    this->fe_values_fs.get_function_values(current_solution,
+                                           this->present_phase_values);
+    this->fe_values_fs.get_function_gradients(current_solution,
+                                              this->phase_gradients);
+    this->fe_values_fs.get_function_laplacians(current_solution,
+                                               this->phase_laplacians);
+
+
+    // Gather previous fs values
+    for (unsigned int p = 0; p < previous_solutions.size(); ++p)
+      {
+        this->fe_values_fs.get_function_values(previous_solutions[p],
+                                               previous_phase_values[p]);
+      }
+
+    // Gather fs stages
+    for (unsigned int s = 0; s < solution_stages.size(); ++s)
+      {
+        this->fe_values_fs.get_function_values(solution_stages[s],
+                                               stages_phase_values[s]);
+      }
+
+
+    for (unsigned int q = 0; q < n_q_points; ++q)
+      {
+        this->JxW[q] = this->fe_values_fs.JxW(q);
+
+        for (unsigned int k = 0; k < n_dofs; ++k)
+          {
+            // Shape function
+            this->phi[q][k]           = this->fe_values_fs.shape_value(k, q);
+            this->grad_phi[q][k]      = this->fe_values_fs.shape_grad(k, q);
+            this->hess_phi[q][k]      = this->fe_values_fs.shape_hessian(k, q);
+            this->laplacian_phi[q][k] = trace(this->hess_phi[q][k]);
+          }
+      }
+  }
+
+  template <typename VectorType>
+  void
+  reinit_velocity(const typename DoFHandler<dim>::active_cell_iterator &cell,
+                  const VectorType &current_solution)
+  {
+    this->fe_values_navier_stokes.reinit(cell);
+
+    this->fe_values_navier_stokes[velocities].get_function_values(
+      current_solution, velocity_values);
+  }
+
+  template <typename VectorType>
+  void
+  reinit_velocity_gradient(const VectorType &current_solution)
+  {
+    this->fe_values_navier_stokes[velocities].get_function_gradients(
+      current_solution, velocity_gradient_values);
+  }
```

Comment:
Couldnt the reinit_velocity reinit everything related to velocity and that's it?

Reply:
done